### DISCUSSION TITLE
Loading supercompressed ktx files

### DISCUSSION THREAD

**Discussion by symBroseph** [vsg-dev/VulkanSceneGraph#1163] [Q&A] [answered]
## tl;dr
Is loading supercompressed ktx files supported? (ie basis-lz format or uastc transcodable format)

If so, how do I do it?

If not, how easy would it be to implement myself?

## How I got here

Hi all,

I'm attempting to load in a supercompressed ktx cubemap as a skybox with the vsgExamples vsgskybox.cpp.

Using the official KTX tool `ktx create` (latest version), I managed to create a working non-compressed cubemap that works just fine.

```
ktx create --format R8G8B8A8_SRGB --cubemap face_rt.png face_lf.png face_up.png face_dn.png face_ft.png face_bk.png cubemap.ktx
```

However, when I encode it with `--encode basis-lz` or `--encode uastc` (the only two options), then I get this error upon loading it into the vsgskybox example:

```
ktx::read(./Assets/Textures/Skyboxes\celestial_grid_cubemap.ktx) failed : Mismatched ktxFormatSize.blockSize and ktxTexture_GetElementSize(texture).
```

I delved into ktx.cpp in the vsgXchange repo and found this section:

```
      const auto format = ktxTexture_GetVkFormat(texture);
      
      ktxFormatSize formatSize;
      vkGetFormatSize(format, &formatSize);
      
      if (formatSize.blockSizeInBits != texture->_protected->_formatSize.blockSizeInBits)
      {
#if 0
            auto before_valueSize = ktxTexture_GetElementSize(texture);

            vkGetFormatSize( format, &(texture->_protected->_formatSize) );

            auto after_valueSize = ktxTexture_GetElementSize(texture);

            std::cout<<"ktx::read("<<filename<<") Fallback : format = "<<format<<", before_valueSize = "<<before_valueSize<<" after_valueSize = "<<after_valueSize<<std::endl;
#endif
            throw vsg::Exception{"Mismatched ktxFormatSize.blockSize and ktxTexture_GetElementSize(texture)."};
      }
```

Which tells me that the vkFormat of the texture is to blame for the mismatch.

Using `ktx info` I can see the format of the encoded cubemap is `VK_FORMAT_UNDEFINED` while the format for the unencoded cubemap is `VK_FORMAT_R8G8B8A8_SRGB`. 

This appears to be intended behavior, as per the [ktx spec](https://github.khronos.org/KTX-Specification/ktxspec.v2.html):

![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/81342925/e7092ac3-c281-4c67-be52-f9032e316e73)

##

So my question is (finally), since this is intended behavior for ktx files, does VSG have a way to deal with the `VK_FORMAT_UNDEFINED` format and load in these textures, or is that not implemented?

If it's not implemented, is there an easy way to bring this functionality into my project?

**Comment by robertosfield** [comment]
I haven't touched the ktx plugin since I originally worked on it.  I'm afraid it's too long for me remember details - you will probably be more informed and up to date by now.

If a value needs to be passed  back as meta data then you can assign a user value by setting data->setValue("property", value);

What you should do with that I can't say at this point, or even whether that's appropriate.

**Comment by robertosfield** [comment, marked as answer]
Thinking back to original work on the ktx plugin (3 years ago) I integrated the ktx source into vsgXchange as fallback because libktx project was painful to download and build at the time.  Things may have changed now.  

I don't think basis support is supported in the local build, it was very early life in it's life back 3 years ago so wasn't a settled lib you could just link to.  If libktx and basis have become more friendly for pulling is an external dependency then it might be time to rewrite the plugin to use an external libktx and basis.  Then figure out full basis support.

I have a lot of work on my plate right now so I'll not be able to look at this refactor, so others will need to be pick up the ball and run with it.

**Reply by symBroseph** [reply]
Understandable, it didn't look like it was an option just from skimming the source code, but wanted to check. Thanks for the quick response!

**Reply by xarray** [reply]
I've integrated the KTX support and BasisU texture encoder/decoder in my osgverse project last year and it works well with osgb tiles, reducing 20-30% size of the original dataset. A reference could be found at:
https://github.com/xarray/osgverse/blob/master/readerwriter/LoadTextureKTX.cpp

I'm going to try merging some of my previous work (including gltf, ktx and lightweighted fbx plugins) to vsg in May and June, after I finish my deadlines on some projects~

**Reply by robertosfield** [reply]
@xarray Draco and KTX/BasisU are two elements that should help with file loading, especially for paged databases so kudos for adding them osgverse.  How are you managing these dependencies?

**Comment by xarray** [comment]
> @xarray Draco and KTX/BasisU are two elements that should help with file loading, especially for paged databases so kudos for adding them osgverse. How are you managing these dependencies?

Draco could be added as an external library and configured in cmake options. For KTX, I include the core source code in https://github.com/xarray/osgverse/tree/master/3rdparty/ktx and keep its own licenses.

**Reply by robertosfield** [reply]
Thanks for the link. I see that you include BasisU as well in your 3rd patry.  This might be a fallback but I'd prefer that these 3rd party dependencies come as easy to use 3rd party libraries.  

Do you know if the situation with building/sourcing them as libs has improved?

**Reply by xarray** [reply]
I've clone the latest code from https://github.com/KhronosGroup/KTX-Software
Build it with CMake under Windows, everything seems to work like a charm now.
